packages/grafana-runtime/src/utils/migrationHandler.ts

```diff
@@ -0,0 +1,70 @@
+import { DataQueryRequest } from '@grafana/data';
+import { DataQuery } from '@grafana/schema';
+
+import { config } from '../config';
+import { getBackendSrv } from '../services';
+
+export interface MigrationHandler {
+  hasBackendMigration: boolean;
+  shouldMigrate(query: DataQuery): boolean;
+}
+
+export function instanceOfMigrationHandler(object: any): object is MigrationHandler {
+  return 'hasBackendMigration' in object && 'shouldMigrate' in object;
```

Comment:
Just to be sure this is not abused by FE datasources, please add `object instanceof DataSourceWithBackend && ..`